File: quotes.py
import random
import os
import sys
import json

# New version rewritten from the ground up
# Ensures that there are no duplicates by creating a new list of quotes and deleting quotes from that list after the script has been run
# Doesn't touch the original quotes file, it copies the contents to the temp file after shuffling them, then reads from the temp file and deletes quotes from that
# Once the temp file is empty the script will delete it, then on the next run it creates a new temp file and starts again

# Temp file must be written to disk for persistance between runs
quotes_file = "quotes.txt"
temp_file = "temp_quotes.txt"

def get_quote():
    if os.path.exists(temp_file):
        with open(temp_file, 'r') as f:
            lines = json.load(f)
    else:
        with open(quotes_file, 'r') as f:
            lines = f.readlines()
        random.shuffle(lines)

    # Pick a quote to print then delete it from temp_quotes.txt
    quote = lines.pop()

    # If the list is empty delete temp file, else write the remaining quotes (minus the one just used) back
    if not lines:
        os.remove(temp_file)
    else:
        with open(temp_file, 'w') as f:
            json.dump(lines, f)

    # The final product
    return quote

# Check if quotes.txt exists before running the function, warn user if not
if not os.path.exists(quotes_file):
    sys.exit(f"quotes.txt is missing - please put a file named quotes.txt in the same directory as this script.")

print(get_quote())

File: quotes.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"math/rand"
	"os"
	"time"
)

func main() {
	rand.Seed(time.Now().UnixNano()) // Generate entropy

	quotesFile := "quotes.txt"
	tempFile := "temp_quotes.txt"

	// Error handling if quotes.txt is missing
	if _, err := os.Stat(quotesFile); os.IsNotExist(err) {
		// quotes.txt does not exist in current directory, tell the user
  		fmt.Println("quotes.txt is missing - please put a file named quotes.txt in the same directory as this script.")
  		return
	}

	var lines []string

	// Check if temp file exists already
	if _, err := os.Stat(tempFile); err == nil {
		// If it does, load quotes from it
		data, _ := ioutil.ReadFile(tempFile)
		json.Unmarshal(data, &lines)
	} else if os.IsNotExist(err) {
		// If not, copy from the original quotes.txt file
		file, _ := os.Open(quotesFile)
		scanner := bufio.NewScanner(file)
		for scanner.Scan() {
			lines = append(lines, scanner.Text())
		}
		file.Close()

		// Mix it up like codeine and Sprite
		rand.Shuffle(len(lines), func(i, j int) { lines[i], lines[j] = lines[j], lines[i] })
	} else {
		fmt.Println("Error checking file:", err)
		return
	}

	// Pick a random quote
	quote := lines[len(lines)-1]
	lines = lines[:len(lines)-1]

	// If the temp file is empty, delete it so it gets recreated on next run
	if len(lines) == 0 {
		os.Remove(tempFile)
	// If not empty, copy remaining quotes back to it for next run
	} else {
		data, _ := json.Marshal(lines)
		ioutil.WriteFile(tempFile, data, 0644)
	}
	
	// Finally, print a quote
	fmt.Println(quote)
}
